File: sort/insertion_sort.rb
# 插入排序
# 对于没有排序的元素，在已经排序的序列里从后向前扫描，然后在相应的位置插入。
# 复杂度：
# 时间复杂度：最坏情况下 O(n2)O(n2)，最好情况下 O(n)O(n)。
# 空间复杂度：O(1)O(1)。
def insertion_sort arr
  n = arr.length
  n.times do |i|
    key = arr[i]
    j = i-1
    while j >=0 and arr[j] > key
      arr[j+1] = arr[j]
      j-=1
    end
    arr[j+1] = key
  end
  arr
end

arr = [4, 2, 1, 5, 29, 25, 22, 29]
insertion_sort arr
puts arr

File: sort/insertion_sort.go
package main

import "fmt"

func insertionSort(arr []int) {
  n := len(arr)
  for i:=0;i<n;i++{
    key := arr[i]
    j := i-1
    for j >= 0 && arr[j] > key {
      arr[j+1] = arr[j]
      j-=1
    }
    arr[j+1] = key
  }
}

func main() {
  arr := []int{5, 4, 2, 1, 7, 22, 21, 26}
  insertionSort(arr)
  fmt.Println(arr)
}
